pgjdbc/src/main/java/org/postgresql/jdbc/ArrayDecoding.java

```diff
@@ -0,0 +1,791 @@
+/*
+ * Copyright (c) 2020, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.jdbc;
+
+import org.postgresql.core.BaseConnection;
+import org.postgresql.core.Oid;
+import org.postgresql.jdbc2.ArrayAssistant;
+import org.postgresql.jdbc2.ArrayAssistantRegistry;
+import org.postgresql.util.GT;
+import org.postgresql.util.PGbytea;
+import org.postgresql.util.PSQLException;
+import org.postgresql.util.PSQLState;
+
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.lang.reflect.Array;
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Utility for decoding arrays.
+ *
+ * <p>
+ * See {@code ArrayEncoding} for description of the binary format of arrays.
+ * </p>
+ *
+ * @author Brett Okken
+ */
+final class ArrayDecoding {
+
+  /**
+   * Array list implementation specific for storing PG array elements. If
+   * {@link PgArrayList#dimensionsCount} is {@code 1}, the contents will be
+   * {@link String}. For all larger <i>dimensionsCount</i>, the values will be
+   * {@link PgArrayList} instances.
+   */
+  static final class PgArrayList extends ArrayList<@Nullable Object> {
+
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * How many dimensions.
+     */
+    int dimensionsCount = 1;
+
+  }
+
+  private interface ArrayDecoder<@NonNull A> {
```

Comment:
Have you tried `ArrayDecoder<A extends @NonNull Object>` here? (I assume you intend that A should be `non-nullable`)

https://checkerframework.org/manual/#inherited-wildcard-annotations